clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2857,32 +2857,63 @@ static void handleWeakImportAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
 }
 
 // Checks correctness of mutual usage of different work_group_size attributes:
-// reqd_work_group_size, max_work_group_size. Values of reqd_work_group_size
-// arguments shall be equal or less than values coming from max_work_group_size.
+// reqd_work_group_size, max_work_group_size and max_global_work_dim.
+// Values of reqd_work_group_size arguments shall be equal or less than values
+// coming from max_work_group_size.
+// In case of 'max_global_work_dim' attribute equals to 0 we shall ensure
+// that if max_work_group_size and reqd_work_group_size attributes exist,
+// then they are equal (1, 1, 1).
 static bool checkWorkGroupSizeValues(Sema &S, Decl *D, const ParsedAttr &Attr,
                                      uint32_t WGSize[3]) {
-  if (const SYCLIntelMaxWorkGroupSizeAttr *A =
-      D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
-    if (!(WGSize[0] <= A->getXDim() && WGSize[1] <= A->getYDim() &&
-          WGSize[2] <= A->getZDim())) {
-      S.Diag(Attr.getLoc(), diag::err_conflicting_sycl_function_attributes)
-          << Attr << A->getSpelling();
-      D->setInvalidDecl();
-      return false;
+  bool Result = true;
+  if (Attr.getKind() == ParsedAttr::AT_SYCLIntelMaxGlobalWorkDim) {
+    if (const SYCLIntelMaxWorkGroupSizeAttr *A =
+            D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+      if (A->getXDim() != 1 || A->getYDim() != 1 || A->getZDim() != 1) {
```

Comment:
I have a suggestion to move this if-condition and diagnostics to a separate function as it's used three times here. However it's up to you

Reply:
Don't see much value in this for now, since actually it's not used 3 times but only 2 (lines 2887-2896 is a different case). Yet I could create a function that checks for WG sizes, where WG would be passed as function arguments (to merge all 3 cases), but in this case it would be a function with multiple arguments which isn't looking good for my perspective.